beluga/include/beluga/actions/assign.hpp

```diff
@@ -0,0 +1,116 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ACTIONS_ASSIGN_HPP
+#define BELUGA_ACTIONS_ASSIGN_HPP
+
+#include <range/v3/action/action.hpp>
+#include <range/v3/functional/bind_back.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/view.hpp>
+
+namespace beluga::actions {
+
+/// \cond
+
+template <class T, class = void>
+struct is_range_closure : std::false_type {};
+
+template <class T>
+struct is_range_closure<ranges::actions::action_closure<T>> : std::true_type {};
+
+template <class T>
+struct is_range_closure<ranges::views::view_closure<T>> : std::true_type {};
+
+template <class T>
+inline constexpr bool is_range_closure_v = is_range_closure<T>::value;
+
+/// \endcond
+
+namespace detail {
+
+/// Implementation detail for an assign range adaptor object.
+struct assign_fn {
+  /// Overload that implements the assign algorithm.
+  template <
+      class Range,
+      class Fn,
+      std::enable_if_t<ranges::range<Range>, int> = 0,
+      std::enable_if_t<is_range_closure_v<Fn>, int> = 0>
+  constexpr auto operator()(Range& range, Fn fn) const -> Range& {
+    auto&& view = fn(range);
+    if constexpr (!std::is_same_v<Range, std::decay_t<decltype(view)>>) {
+      // If the result of invoking the closure is not the range itself,
+      // then we need to convert the view and assign it to the input range.
+      range = std::move(view) | ranges::to<Range>;
```

Comment:
@nahueespinosa nit: does an rvalue `view` force move-construction of `Range` elements? Or do we need [`ranges::views::move`](https://ericniebler.github.io/range-v3/structranges_1_1views_1_1move__fn.html)?

Reply:
That's a great question, here we move the view itself, not the elements. I think copying the elements by default is the sensible choice, since we don't know if the view will require access to the same element more than once. Our own sample view would be one such case. This is also what [`std::ranges::to`](https://en.cppreference.com/w/cpp/ranges/to#Notes) does by default too.

I added documentation explaining this and a test showing how to move in the elements in https://github.com/Ekumen-OS/beluga/pull/287/commits/905fcd01f634bdbc0b5c2d9658f7fc85b7d84f73.